### DISCUSSION TITLE
In depth knowledge of Meta-Structures and variables

### DISCUSSION THREAD

**Discussion by sabyasachi07** [Blaze-Stars/Image-Filter-Application#41] [Q&A] [answered]
![part_1](https://user-images.githubusercontent.com/55123454/171261330-ff257d6e-1448-4ca9-ace9-9cd39712ab42.jpg)
![part_2](https://user-images.githubusercontent.com/55123454/171261344-76a35238-6153-4ee0-a12b-bd298612343f.jpg)
![part_3](https://user-images.githubusercontent.com/55123454/171261361-54b6fce9-f6c8-47a2-bd5d-6040fd9a6501.jpg)

**Comment by chandrakant100** [comment]
Good job @sabyasachi07 it will be really helpful for us while writing code in `main.cpp`

**Comment by H1manshu21** [comment, marked as answer]
Actual calculation for `bfOffBits` of `BITMAPFILEHEADER`,
```cpp
bfOffBits = sizeof (BITMAPFILEHEADER) + NumberOfRGBTRIPLE * sizeof (RGBTRIPLE)  + sizeof (BITMAPINFOHEADER);
```
So looking into our 24bit image, it looks like this:
```cpp
bfOffBits = 14 + 0 * 4  + 40;
// Prints 54
std::cout << bfOffBits;
```

- `RGBTRIPLE` variable is a `color table`. It is commonly either `16 colors or 256 colors`, depending on how the file is specified in `BITMAPINFOHEADER’s biBitCount`.

- In our case, it does not exist in the file because we are using a `24-bit bitmap` which is `2^24 = 16777216` colors and doesn't fit in `RGBTRIPLE` and the `bmiColors` member of [BITMAPINFO](https://msdn.microsoft.com/en-us/library/dd183375(v=vs.85)) is NULL.
- Each `3-byte triplet` in the bitmap array represents the relative intensities of `blue, green, and red`, respectively, for a pixel. The `bmiColors` color table is used for optimizing colors used on `palette-based devices`, and must contain the number of entries specified by the `biClrUsed` member of the `BITMAPINFOHEADER`.

**SOURCE**: https://chowdera.com/2022/04/202204170049352833.html

**Reply by sabyasachi07** [reply]
Do u have any source or link where we can get brief details to support your calculation, As I have a different calculation regarding this.

**Reply by sabyasachi07** [reply]
![AllBMPformats](https://user-images.githubusercontent.com/55123454/171344349-4f2cb418-4a8a-485e-851d-36bdaa5954a0.png)
From picture u can see 1st row is all header i.e bitmapinfoheader and bitmapfileheader. After bitmapinfoheader real image data comes so you can't separately add `NumberOfRGBQUAD * sizeof (RGBQUAD)` with bitmapfileheader.

**Reply by chandrakant100** [reply]
`RGBTRIPLE` and `RGBQUAD` are two separate structures Since we are working with 24-bit bitmaps not 32-bit, there is no need to use `RGBQUAD` to store extra reserve information.

**Reply by H1manshu21** [reply]
`rgbReserved` is reserved and must be zero according to Microsoft. I think `RGBTRIPLE` is being used instead of `RGBQUAD` since it is storing relative intensities of blue, green, and red.

**Reply by H1manshu21** [reply]
May be like this 
```cpp
bfOffBits = sizeof (BITMAPFILEHEADER) + NumberOfRGBTRIPLE * sizeof (RGBTRIPLE)  + sizeof (BITMAPINFOHEADER);
```

**Reply by chandrakant100** [reply]
@Himanshu40 should we worry about the internal calculation because it is given by Microsoft

**Reply by H1manshu21** [reply]
> @Himanshu40 should we worry about the internal calculation because it is given by Microsoft

What about this?
```cpp
bfOffBits = sizeof (BITMAPFILEHEADER) + NumberOfRGBTRIPLE * sizeof (RGBTRIPLE)  + sizeof (BITMAPINFOHEADER);
```

**Reply by chandrakant100** [reply]
This might be the calculation for `bfOffBits`. I am not sure !! 
From the sources above It looks fine actually.

**Reply by chandrakant100** [reply]
> > @Himanshu40 should we worry about the internal calculation because it is given by Microsoft
> 
> What about this?
> 
> ```c++
> bfOffBits = sizeof (BITMAPFILEHEADER) + NumberOfRGBTRIPLE * sizeof (RGBTRIPLE)  + sizeof (BITMAPINFOHEADER);
> ```

Have you tested it through `std::cout` @Himanshu40 ?

**Reply by H1manshu21** [reply]
> > > @Himanshu40 should we worry about the internal calculation because it is given by Microsoft
> > 
> > 
> > What about this?
> > ```c++
> > bfOffBits = sizeof (BITMAPFILEHEADER) + NumberOfRGBTRIPLE * sizeof (RGBTRIPLE)  + sizeof (BITMAPINFOHEADER);
> > ```
> 
> Have you tested it through `std::cout` @Himanshu40 ?

Check: https://chowdera.com/2022/04/202204170049352833.html

**Reply by chandrakant100** [reply]
The calculation looks fine!

**Reply by H1manshu21** [reply]
> The calculation looks fine!

Since in `bmp.h` there are three classes and they are being calculated as one I think? `RGBTRIPLE` structure describes a color consisting of relative intensities of red, green, and blue.

**Reply by sabyasachi07** [reply]
https://topic.alibabacloud.com/a/bitmapinfoheader-and-bitmapfileheader_8_8_32023995.html
